File: mahjong_autocalc/vision/pipeline.py
"""画像 → 牌の並び、までを通す。"""

from __future__ import annotations

import base64
from dataclasses import dataclass

import cv2
import numpy as np

from ..tiles import tile_name, tile_to_str
from . import heuristic
from .detect import DetectedTile, DetectionError, detect_tiles
from .features import TileFeatures, extract
from .library import TileLibrary
from .prior import load_prior_library

# 照合スコアがこの値を超えたら採用する。順に「自分で登録した牌 → 同梱の初期
# ライブラリ → 規則ベース」と落とす。
#
# しきい値を下げて弱い一致まで拾うと、規則ベースなら当たっていた牌を潰して
# しまい、かえって正解率が下がる (実測で 100% → 93%)。低い一致度は採用せず、
# 次の段に譲る。
LIBRARY_TRUST_THRESHOLD = 0.86
PRIOR_TRUST_THRESHOLD = 0.80
LOW_CONFIDENCE = 0.55


@dataclass(frozen=True)
class TileGuess:
    index: int
    tile: int | None
    confidence: float
    source: str  # "library" | "heuristic" | "unknown"
    group: int
    alternatives: tuple[tuple[int, float], ...]
    crop_png: bytes

    @property
    def is_uncertain(self) -> bool:
        return self.tile is None or self.confidence < LOW_CONFIDENCE

    def to_json(self) -> dict:
        return {
            "index": self.index,
            "tile": self.tile,
            "name": tile_name(self.tile) if self.tile is not None else None,
            "code": tile_to_str(self.tile) if self.tile is not None else None,
            "confidence": round(self.confidence, 3),
            "source": self.source,
            "group": self.group,
            "uncertain": self.is_uncertain,
            "alternatives": [
                {"tile": t, "name": tile_name(t), "score": round(s, 3)}
                for t, s in self.alternatives
            ],
            "crop": "data:image/png;base64," + base64.b64encode(self.crop_png).decode(),
        }


@dataclass(frozen=True)
class RecognitionResult:
    guesses: tuple[TileGuess, ...]
    library_size: int

    @property
    def tiles(self) -> list[int | None]:
        return [g.tile for g in self.guesses]

    @property
    def uncertain_count(self) -> int:
        return sum(1 for g in self.guesses if g.is_uncertain)

    def to_json(self) -> dict:
        return {
            "tiles": [g.to_json() for g in self.guesses],
            "count": len(self.guesses),
            "uncertain_count": self.uncertain_count,
            "library_size": self.library_size,
        }


def _encode_png(image: np.ndarray) -> bytes:
    ok, buffer = cv2.imencode(".png", image)
    if not ok:
        return b""
    return buffer.tobytes()


def decode_image(data: bytes) -> np.ndarray:
    array = np.frombuffer(data, dtype=np.uint8)
    image = cv2.imdecode(array, cv2.IMREAD_COLOR)
    if image is None:
        raise DetectionError("画像を読み込めませんでした (対応形式: JPEG / PNG / WebP)")
    return image


def classify_one(
    features: TileFeatures,
    library: TileLibrary | None,
    prior: TileLibrary | None = None,
) -> tuple[int | None, float, str, list[tuple[int, float]]]:
    """1 枚ぶんの判定。

    優先順位は「自分で登録した牌 → 同梱の初期ライブラリ → 規則ベース」。
    """
    own = library.match(features) if library is not None else None
    base = prior.match(features) if prior is not None else None
    rules = heuristic.classify(features)

    def accept(match, source):
        # 一致度が高くても 2 位と僅差なら確信度を下げる。
        confidence = min(0.99, match.score * (0.75 + min(match.margin, 0.2)))
        return match.tile, confidence, source, [(match.tile, match.score)] + rules[:3]

    if own is not None and own.score >= LIBRARY_TRUST_THRESHOLD:
        return accept(own, "library")
    if base is not None and base.score >= PRIOR_TRUST_THRESHOLD:
        return accept(base, "prior")

    if rules:
        tile, score = rules[0]
        return tile, score, "heuristic", rules[:4]

    fallback = own or base
    if fallback is not None:
        source = "library" if own is not None else "prior"
        return fallback.tile, fallback.score * 0.5, source, [(fallback.tile, fallback.score)]

    return None, 0.0, "unknown", []


def recognize(
    image: np.ndarray,
    library: TileLibrary | None = None,
    prior: TileLibrary | None | str = "default",
) -> RecognitionResult:
    """写真から牌を認識する。

    ``prior`` を省略すると同梱の初期ライブラリを使う。``None`` を渡すと使わない。
    """
    if prior == "default":
        prior = load_prior_library()
    detected: list[DetectedTile] = detect_tiles(image)

    guesses: list[TileGuess] = []
    for tile_image in detected:
        features = extract(tile_image.image)
        tile, confidence, source, alternatives = classify_one(features, library, prior)
        guesses.append(
            TileGuess(
                index=tile_image.order,
                tile=tile,
                confidence=float(confidence),
                source=source,
                group=tile_image.group,
                alternatives=tuple(alternatives),
                crop_png=_encode_png(tile_image.image),
            )
        )

    return RecognitionResult(
        tuple(guesses), library.size if library is not None else 0
    )


def learn_from_corrections(
    image: np.ndarray, assignments: dict[int, int], library: TileLibrary
) -> int:
    """認識結果の手直しをライブラリに反映する。

    Args:
        image: 認識に使ったのと同じ画像。
        assignments: 検出インデックス → 正しい牌 ID。
        library: 追記先。

    Returns:
        実際に登録した枚数。すでに覚えている見た目と区別がつかないものは
        数えない (同じ写真を繰り返し登録してもライブラリが太らないため)。
    """
    detected = detect_tiles(image)
    by_index = {t.order: t for t in detected}
    learned = 0
    for index, tile in assignments.items():
        source = by_index.get(index)
        if source is None:
            continue
        if library.add(tile, extract(source.image)):
            learned += 1
    if learned:
        library.save()
    return learned


__all__ = [
    "LOW_CONFIDENCE",
    "RecognitionResult",
    "TileGuess",
    "decode_image",
    "learn_from_corrections",
    "recognize",
]

File: webapp/vision/pipeline.js
// 画像 → 牌の並び、までを通す。Python 版 vision/pipeline.py の移植。

import { tileName, tileToStr } from "../engine/tiles.js";
import { MAX_DIMENSION, cropRow, detectTiles } from "./detect.js";
import { extract } from "./features.js";
import { classify } from "./heuristic.js";
import { resize } from "./cv.js";

// 照合スコアがこの値を超えたら採用する。順に「自分で登録した牌 → 同梱の初期
// ライブラリ → 規則ベース」と落とす。
//
// しきい値を下げて弱い一致まで拾うと、規則ベースなら当たっていた牌を潰して
// しまい、かえって正解率が下がる (実測で 100% → 93%)。低い一致度は採用せず、
// 次の段に譲る。
export const LIBRARY_TRUST_THRESHOLD = 0.86;
export const PRIOR_TRUST_THRESHOLD = 0.80;
export const LOW_CONFIDENCE = 0.55;

/**
 * 1 枚ぶんの判定。
 *
 * 優先順位は「自分で登録した牌 → 同梱の初期ライブラリ → 規則ベース」。
 * 手元のセットを登録してあれば一番強いが、無くても初期ライブラリで大半は読める。
 */
export function classifyOne(features, library, prior = null) {
  const own = library ? library.match(features) : null;
  const base = prior ? prior.match(features) : null;
  const rules = classify(features);

  const accept = (match, source) => ({
    tile: match.tile,
    // 一致度が高くても 2 位と僅差なら確信度を下げる。
    confidence: Math.min(0.99, match.score * (0.75 + Math.min(match.margin, 0.2))),
    source,
    alternatives: [[match.tile, match.score], ...rules.slice(0, 3)],
  });

  if (own && own.score >= LIBRARY_TRUST_THRESHOLD) return accept(own, "library");
  if (base && base.score >= PRIOR_TRUST_THRESHOLD) return accept(base, "prior");

  if (rules.length) {
    return { tile: rules[0][0], confidence: rules[0][1], source: "heuristic", alternatives: rules.slice(0, 4) };
  }

  const fallback = own || base;
  if (fallback) {
    return {
      tile: fallback.tile,
      confidence: fallback.score * 0.5,
      source: own ? "library" : "prior",
      alternatives: [[fallback.tile, fallback.score]],
    };
  }

  return { tile: null, confidence: 0, source: "unknown", alternatives: [] };
}

function describe(detected, library, prior) {
  return detected.map((tile) => {
    const features = extract(tile.image);
    const { tile: id, confidence, source, alternatives } = classifyOne(features, library, prior);
    return {
      index: tile.order,
      tile: id,
      name: id === null ? null : tileName(id),
      code: id === null ? null : tileToStr(id),
      confidence,
      source,
      group: tile.group,
      uncertain: id === null || confidence < LOW_CONFIDENCE,
      alternatives: alternatives.map(([t, s]) => ({ tile: t, name: tileName(t), score: s })),
      face: tile.image,
      quad: tile.quad,
      features,
    };
  });
}

function summarize(guesses, library) {
  return {
    guesses,
    count: guesses.length,
    uncertainCount: guesses.filter((g) => g.uncertain).length,
    librarySize: library ? library.size : 0,
  };
}

/**
 * 写真から牌を認識する。
 * @param {object} image RGB 画像
 * @param {TileLibrary|null} library
 */
export function recognize(image, library = null, prior = null) {
  return summarize(describe(detectTiles(image), library, prior), library);
}

/**
 * 利用者が指定した範囲だけを読む。
 *
 * 自動検出は背景や照明の影響を受けるが、こちらは範囲と枚数が与えられるので
 * そこは外さない。切り出しと判別は自動検出と同じものを通す。
 *
 * @param {object} image RGB 画像 (recognize に渡すのと同じもの)
 * @param {{quad:number[][], count:number}[]} regions 画像座標での範囲と枚数
 */
export function recognizeRegions(image, regions, library = null, prior = null) {
  // 自動検出と同じ土俵に乗せるため、同じ縮小をかけてから切り出す。
  let work = image;
  let scale = 1;
  const longest = Math.max(image.width, image.height);
  if (longest > MAX_DIMENSION) {
    scale = MAX_DIMENSION / longest;
    work = resize(image, Math.round(image.width * scale), Math.round(image.height * scale));
  }
  const invScale = 1 / scale;

  const detected = [];
  regions.forEach((region, group) => {
    const quad = region.quad.map(([x, y]) => [x * scale, y * scale]);
    const count = Math.max(1, Math.min(18, Math.round(region.count)));
    for (const tile of cropRow(work, quad, count, group, detected.length)) {
      detected.push({ ...tile, quad: tile.quad.map(([x, y]) => [x * invScale, y * invScale]) });
    }
  });

  return summarize(describe(detected, library, prior), library);
}

/**
 * 認識結果の手直しをライブラリに反映する。
 * @param {object[]} guesses recognize() の結果 (features を持っているもの)
 * @param {Map<number,number>|object} assignments 検出インデックス → 正しい牌 ID
 * @returns {number} 実際に登録した枚数
 */
export function learnFromCorrections(guesses, assignments, library) {
  const lookup = assignments instanceof Map ? assignments : new Map(
    Object.entries(assignments).map(([k, v]) => [Number(k), v])
  );

  let learned = 0;
  for (const guess of guesses) {
    const tile = lookup.get(guess.index);
    if (tile === undefined || tile === null) continue;
    if (library.add(tile, guess.features)) learned += 1;
  }
  return learned;
}
